---
title: "RPM K5 vs CGRP preprocessing"
output: html_notebook
date: 2025-06-17
author: "Abbie Ireland, Darren Tyson"
---

## Preprocessing data
Primarily adding signature scores and reformatting for different plotting functions.

```{r}
OVERWRITE <- FALSE
```

```{r}
suppressPackageStartupMessages({ 
    library(tidyverse)
    library(SingleCellExperiment)
    # library(SummarizedExperiment)
    library(ggplot2)
    library(cowplot)
    library(Seurat)
    library(viridis)
    library(ggpubr)
    library(anndataR)
    library(dplyr)
    library(FSA)       # For dunnTest
    library(rstatix)   # Optional, nice formatting of results
    library(patchwork)
    library(gridExtra)
})

```

### Read in AnnData object and convert to Seurat object
AnnData object generated by ?? (location of processing code)
```{r}
adata <- anndataR::read_h5ad("../data/021825_RPM_CGRPvK5_adata3.h5ad")
# Convert AnnData to SingleCellExperiment
sce <- adata$as_SingleCellExperiment()

RPM_K5vCGRP <- CreateSeuratObject(counts = counts(sce), meta.data = as.data.frame(colData(sce)))

#Create an assay of normalized gene expression
RPM_K5vCGRP[["norm"]] <- CreateAssayObject(counts = assay(sce, "norm"))

# Set assay norm as default assay for seurat object
DefaultAssay(RPM_K5vCGRP) <- 'norm'

rm('sce')
```

### Show contents of `RPM_K5vCGRP`
```{r}
RPM_K5vCGRP
```

### Add metadata from AnnData object to Seurat object
```{r}
# Add embeddings of umap, and X_scVI_1.3 to assay norm
temp <- adata$obsm$X_umap
colnames(temp)<-c("UMAP_1","UMAP_2")
rownames(temp)<-colnames(RPM_K5vCGRP)

RPM_K5vCGRP[['umap']] <- CreateDimReducObject(temp, key="UMAP_", assay = "norm") # NOTE: this was 'RNA' in original code, even though comment above says 'norm'
head(RPM_K5vCGRP[['umap']]@cell.embeddings)
```
```{r Set Idents to Cre}
Idents(RPM_K5vCGRP) <- 'Cre'
```

```{r Normalize expression data}
# NOTE: norm assay already exists; need again?
DefaultAssay(RPM_K5vCGRP) <- 'RNA'
RPM_K5vCGRP <- NormalizeData(RPM_K5vCGRP)
```

### Gene signature scores
For Fig. 1k and Extended Data Fig. 2g

```{r Get gene signatures}
# mouse-human orthologs
mouse94 <- read.csv("../data/Signatures/mouse94.csv", row.names = 1)

# human gene signatures for archetypes
arch <- read.csv("../data/Signatures/Archetype_Sigs_Maddox.csv")

```


```{r Calculate archetype scores}
a <- arch$SCLC.A
a2 <- arch$SCLC.A2
n <- arch$SCLC.N
p <- arch$SCLC.P
y <- arch$SCLC.Y

a <- a[1:987]
a_sc <- subset(mouse94, mouse94$human_homolog %in% a)
a_sc_mouse <- (a_sc$gene_name)

n_sc <- subset(mouse94, mouse94$human_homolog %in% n)
n_sc_mouse <- (n_sc$gene_name)

a2_sc <- subset(mouse94, mouse94$human_homolog %in% a2)
a2_sc_mouse <- (a2_sc$gene_name)

p_sc <- subset(mouse94, mouse94$human_homolog %in% p)
p_sc_mouse <- (p_sc$gene_name)

y_sc <- subset(mouse94, mouse94$human_homolog %in% y)
y_sc_mouse <- (y_sc$gene_name)

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(a_sc_mouse),
    name = 'A_Archetype')

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(a2_sc_mouse),
    name = 'A2_Archetype')

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(n_sc_mouse),
    name = 'N_Archetype')

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(p_sc_mouse),
    name = 'P_Archetype')

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(y_sc_mouse),
    name = 'Y_Archetype')
```

# Human A N P scRNA seq sigs from Chan et al 
Extended Data Fig. 2h
```{r Chan sigs}
sc_sclc_sigs<-read.csv("../data/Signatures/hSCLC_Chan_sigs.csv")

a_sc <- sc_sclc_sigs$hSCLC_A[1:67]
a_sc <- subset(mouse94, mouse94$human_homolog %in% a_sc)
a_sc_mouse <- (a_sc$gene_name)

n_sc <- sc_sclc_sigs$hSCLC_N[1:73]
n_sc <- subset(mouse94, mouse94$human_homolog %in% n_sc)
n_sc_mouse <- (n_sc$gene_name)

p_sc <- sc_sclc_sigs$hSCLC_P
p_sc <- subset(mouse94, mouse94$human_homolog %in% p_sc)
p_sc_mouse <- (p_sc$gene_name)

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(a_sc_mouse),
    name = 'hSCLC_A')

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(n_sc_mouse),
    name = 'hSCLC_N')

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(p_sc_mouse),
    name = 'hSCLC_P')
```

### A, N, P ChIP target genes as scores
Extended Data Fig. 1e
```{r}
chip <- read.csv("../data/Signatures/ASCL1_NEUROD1_POU2F3_ChIP_Targets.csv")
achip <- chip$Borromeo_ASCL1_Targets
nchip <- chip$Borromeo_Oliver_NEUROD1_Targets
pchip <- chip$POU2F3_Targets
pchip <- subset(mouse94, mouse94$human_homolog %in% pchip)
pchip_m <- (pchip$gene_name)

RPM_K5vCGRP<-AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(achip),
    name = 'ASCL1_Targets')

RPM_K5vCGRP<-AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(nchip),
    name = 'NEUROD1_Targets')

RPM_K5vCGRP<-AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(pchip_m),
    name = 'POU2F3_Targets')
```


### Look at basal vs luminal hillock cell state sigs in CGRP vs K5 ####
```{r}
hillock <- read.csv("../data/Signatures/luminal_basal_hillock.csv")
basal <- hillock$Basal_hillock[1:170]
lum <- hillock$Luminal_hillock

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(basal),
    name = 'Basal_hillock')

RPM_K5vCGRP <- AddModuleScore(
    object = RPM_K5vCGRP,
    features = list(lum),
    name = 'Luminal_hillock')
```

### NE Score
Ext. Data Fig. 2f  
#### Converting to SCE to Add NE Score
NE score assignment originally describe in Zhang, TLCR, 2017
```{r}
sce <- as.SingleCellExperiment(RPM_K5vCGRP, assay = "norm")

# import the NE/Non-NE data from the literature
nedat_mouse <- read.csv("../data/Signatures/nedat_mouse.csv")
gene_idx <- rownames(sce) %in% nedat_mouse$gene_name
X <- as.matrix(assay(sce, "logcounts")[gene_idx,])
# assure the gene order matches up
X <- X[nedat_mouse$gene_name,]

# define the scoring function based on methods from Zhang et al TLCR 2018
ne_score <- function(cell, ne, notne, ...) { 
  # report missing value is no variation 
  if (sd(cell) == 0) { 
    NA
  } else { 
    (cor(x = cell, y = ne, ...) - cor(x = cell, y = notne, ...))/2
  }
}
# get the NE scores. note that 4 cells have missing values b/c 
# they had zero variation across the 50 genes. 

sce$NE_spearman <- apply(X = X, 2, ne_score,  
                         ne = nedat_mouse$NE, 
                         notne = nedat_mouse$NonNE, 
                         method = "spearman")

#Add NE score as metadata to Seurat object
RPM_K5vCGRP@meta.data$NE_spearman <- sce$NE_spearman
```
```{r}
# Save the Seurat object with the new metadata
Fig1_data_filepath <- "../data/RPM_K5vCGRP_Seurat.Rds"
if (OVERWRITE || !file.exists(Fig1_data_filepath)) {
    message(paste0("Saving data to ",Fig1_data_filepath))
    saveRDS(RPM_K5vCGRP, file = Fig1_data_filepath)
} else {
    message(paste0(Fig1_data_filepath, " already exists. Set OVERWRITE = TRUE to overwrite."))
}
```


